### PR TITLE
Dijkstra's algorithm

### PR DESCRIPTION
**Pull request by TaliBenedek**

### CODE REVIEW

File: src/main/java/Node.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
+import java.util.*;
+
+public class Node implements Comparable<Node>
+{
+    private SubwayStations.Station station;
+    private double distance = Integer.MAX_VALUE; //distance from source node
+    private List<SubwayStations.Station> shortestPath = new LinkedList<>();
```
Comment (andrewoid) [inline src/main/java/Node.java]: Why are you using a LinkedList?

File: src/main/java/SubwayGraph.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
+import java.util.*;
+
+public class SubwayGraph
+{
+    Set<Node> nodes = new HashSet<>();
+    Map<SubwayStations.Station, Node> map = new HashMap<>();
+
+    public void setUpGraph(SubwayStations subway)
```
Comment (andrewoid) [inline src/main/java/SubwayGraph.java]: This should be a constructor.

File: src/test/java/SubwayGraphTest.java [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        SubwayLines lines = linesConverter.readJsonAsMap();
+        subway.connectStations(lines);
+
+        //when
+        graph.setUpGraph(subway);
+
+        //then
+        Iterator it = graph.nodes.iterator();
+        int index = 0;
+        while (it.hasNext())
```
Comment (andrewoid) [inline src/test/java/SubwayGraphTest.java]: Use a for-each loop.

File: src/test/java/SubwayGraphTest.java [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+        SubwayGraph graph = new SubwayGraph();
+        JsonToSubwayStations subwayConverter = new JsonToSubwayStations();
+        SubwayStations subway = subwayConverter.readJsonObject();
+        JsonToSubwayLines linesConverter = new JsonToSubwayLines();
+        SubwayLines lines = linesConverter.readJsonAsMap();
+        subway.connectStations(lines);
+        graph.setUpGraph(subway);
+        SubwayStations.Station start = subway.stations.get(0);
+        SubwayStations.Station destination = subway.stations.get(32);
+        LinkedList<SubwayStations.Station> expected = new LinkedList<>();
```
Comment (andrewoid) [inline src/test/java/SubwayGraphTest.java]: Arrays.asList(...)

File: src/test/java/SubwayGraphTest.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+        expected.add(subway.stations.get(32));
+
+        //when
+        LinkedList<SubwayStations.Station> shortestPath = graph.shortestPath(start, destination);
+
+        //then
+        for (int index = 0; index < shortestPath.size(); index++)
+        {
+            assertEquals(expected.get(index), shortestPath.get(index));
+        }
```
Comment (andrewoid) [inline src/test/java/SubwayGraphTest.java]: `assertEquals(expected, shortestPath)`

File: src/main/java/Node.java [hunk 6]
```diff
@@ -0,0 +1,98 @@
+import java.util.*;
+
+public class Node implements Comparable<Node>
+{
+    private SubwayStations.Station station;
+    private double distance = Integer.MAX_VALUE; //distance from source node
+    private List<SubwayStations.Station> shortestPath = new ArrayList<>();
+    private List<Node> neighbors = new ArrayList<>();
+    private Node previous;
```
Comment (andrewoid) [inline src/main/java/Node.java:9]: I don't know why you need both `previous` and `shortestPath`

File: src/main/java/SubwayGraph.java [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+            }
+        }
+    }
+
+    public List<SubwayStations.Station> shortestPath(SubwayStations.Station source, SubwayStations.Station destination)
+    {
+        Node sourceNode = map.get(source);
+        Node destinationNode = map.get(destination);
+        sourceNode.setDistance(0);
+        List<Node> visitedNodes = new ArrayList<>();
```
Comment (andrewoid) [inline src/main/java/SubwayGraph.java]: Not sure why you need visited nodes if you have unvisited nodes

**Review by andrewoid — APPROVED** [review]

File: src/main/java/Node.java (hunk 6 above)
Comment (TaliBenedek) [reply src/main/java/Node.java:9]: It was a mistake to have both. I originally was using `shortestPath` but then switched to `previous` and just forgot to remove `shortestPath` afterwards.

**Pull request merged**